drivers/samsung/displays/mdc_protocol.cr

```diff
@@ -0,0 +1,389 @@
+    # options[:emit] = block unless block.nil?
+    do_send(Command::Panel_Mute, Bytes.empty, **options)
+  end
+
+  # Mutes audio + video
+  def mute(
+    state : Bool = true,
+    index : Int32 | String = 0,
+    layer : MuteLayer = MuteLayer::AudioVideo
+  )
+    logger.debug { "requested mute state: #{state}" }
+    mute_video(state)
+    mute_audio(state)
+  end
+
+  # Adds mute states compatible with projectors
+  def mute_video(state : Bool = true)
+    state = state ? 1 : 0
+    do_send(Command::Panel_Mute, state)
+  end
+
+  def unmute_video
+    mute_video(false)
+  end
+
+  # Emulate mute
+  def mute_audio(val : Bool = true)
+    if val
+      if !self[:audio_mute]?
+        self[:audio_mute] = true
+        @previous_volume = self[:volume].as_i
+        volume(0)
+      end
+    else
+      unmute_audio
+    end
+  end
+
+  def unmute_audio
+    if self[:audio_mute]?
+      self[:audio_mute] = false
+      volume(@previous_volume)
+    end
+  end
+
+  # check software version
+  def software_version?
+    do_send(Command::Software_Version)
+  end
+
+  def serial_number?
+    do_send(Command::Serial_Number)
+  end
+
+  def switch_to(input : Input, **options)
+    @input_stable = false
+    @input_target = input
+    do_send(Command::Input, input.value, **options)
+  end
+
+  enum SpeakerModes
+    Internal = 0
+    External = 1
+  end
+
+  def speaker_select(mode : String, **options)
+    do_send(Command::Speaker, SpeakerModes.parse(mode).value, **options)
+  end
+
+  def do_poll
+    do_send(Command::Status, Bytes.empty, priority: 0)
+    power? unless self[:hard_off]?
+  end
+
+  # Enable power on (without WOL)
+  def network_standby(enable : Bool, **options)
+    state = enable ? 1 : 0
+    do_send(Command::Net_Standby, state, **options)
+  end
+
+  # Eco auto power off timer
+  def auto_off_timer(enable : Bool, **options)
+    state = enable ? 1 : 0
+    do_send(Command::Eco_Solution, Bytes[0x81, state], **options)
+  end
+
+  # Device auto power control (presumably signal based?)
+  def auto_power(enable : Bool, **options)
+    state = enable ? 1 : 0
+    do_send(Command::Auto_Power, state, **options)
+  end
+
+  # Display control
+  {% for name in ["volume", "contrast", "brightness", "sharpness", "colour", "tint", "red_gain", "green_gain", "blue_gain"] %}
+    @[Security(Level::Administrator)]
+    def {{name.id}}(val : Int32, **options)
+      val = val.clamp(0, 100)
+      do_send(Command.parse({{name.id.stringify}}), val, **options)
+    end
+  {% end %}
+
+  DEVICE_SETTINGS = [
+    :network_standby,
+    :auto_off_timer,
+    :auto_power,
+    :contrast,
+    :brightness,
+    :sharpness,
+    :colour,
+    :tint,
+    :red_gain,
+    :green_gain,
+    :blue_gain,
+]
+
+  def do_device_config
+    logger.debug { "Syncronising device state with settings" }
+    DEVICE_SETTINGS.each do |name|
+      value = setting?(Int32, name)
+      # TODO: find out if these are equivalent
+      # __send__(name, value) unless value.nil?
```

Comment:
```suggestion
```
I believe these are equivalent.

Pinging @stakach for confirmation.

Reply:
They’re not equivalent.

`__send__` is part of Ruby’s dynamic dispatch. It let’s you invoke an arbitrary method based on a name passed at runtime (https://ruby-doc.org/core-2.7.1/BasicObject.html#method-i-__send__). It’s also available as `send` (original implementation), but the double underscore alias was added for cases where an object has an actual object specific `send` method, which in the case of drivers in Engine, it does (https://docs.acaengine.com/developer-guide/drivers/device#transmitting-a-request).

So in the case above, the commented code would invoke the method called `name` and pass it value. As this is pulling from a constant, the equivalent in crystal in this context would be to use a macro:
```crystal
{% for name in DEVICE_SETTINGS %}
  %value = setting?(Int32, {{name.id}})
  {{name.id}}(%value) unless %value.nil?
{% end %}
```

Keep in mind though that crystal != ruby so you can likely come up with a neater implementation to satisfy the same requirements (query possibly defined device config from settings and push to device).

Reply:
ahh the ol dynamic method dispatch

Reply:
Hey @KimBurgess, thanks for that, good to know what that means in ruby!

I'm trying to write this in crystal and following your suggestion have made this commit to split the boolean and integer device settings but I can't seem to get this to work. I'm not sure if it's possible to use a macro within a method but any help would be great.

https://github.com/PlaceOS/drivers/pull/19/commits/f3969c2b978e08884fd9344ac8c3a57ced6674c8

Reply:
I've also tried using `value` instead of `%value` as I'm not sure what `%` actually means syntax wise which doesn't work either.

Reply:
`%value` creates a unique local macro variable. It essentially prefixes the variable by a random string so you don't end up with accidental variable shadowing within the same scope.

Reply:
fixed this, guess I just needed to take a 2nd look at this and stringify the setting names. Can you take another look at this @KimBurgess https://github.com/PlaceOS/drivers/pull/19/commits/5c868f64a3435147e7d698bdb932cd6fdd212072

Reply:
Cool. That’s coming together. With some of these changes it looks to have led to a split with the way you’re handling these settings methods and their application based on the type (`Bool` or `Int32`). Something that may be worth playing with is removing that split by changing the way you’re defining what settings are available.

```crystal
DEVICE_SETTINGS = {
  network_standby: Bool,
  volume: Int32,
}

{% for name, kind in DEVICE_SETTINGS %}
  def {{name.id}}(val : {{kind}})
    {% if kind.resolve == Bool %}
      val = val ? 1 : 0
    {% elsif kind.resolve == Int32 %}
      val = val.clamp 0, 100
    {% end %}
    do_send Command::{{name.id.upcase}}, val
  end
{% end %}

# etc - similar impl for applying these based on PlaceOS settings now that you have the type info available.
```

Reply:
Hmmmm, that’s annoyingly unique. You could either keep that as a standalone method, or add in a special case within that macro. Neither is ideal, but I do think the concept of declaring all settings in one spot is worthwhile.

Reply:
yeah definitely agree on declaring them all in one spot, I've made this commit for that https://github.com/PlaceOS/drivers/pull/19/commits/ad373785f29e399586a1d5613e9d387419fa3855

Reply:
Hey @KimBurgess , can you give this another review when you can please?